batch-indexing/src/main/java/data/config/IndexingJobConfig.java

```diff
@@ -0,0 +1,125 @@
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.task.TaskExecutor;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
+import org.springframework.transaction.PlatformTransactionManager;
+
+import java.lang.invoke.MethodHandles;
+import java.util.concurrent.Future;
+
+import static data.job.BatchJobType.INDEXING;
+
+@Configuration
+@EnableTask
+@ComponentScan(basePackages = "data.repositories")
+public class IndexingJobConfig {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+    public static final String BATCH_JOB = INDEXING.name();
+    public static final String STEP_NAME = "indexingStep";
+
+    @Value("${chunk.size}")
+    public int chunkSize;
+    @Value("${parallelization.size}")
+    public int parallelization;
+
+    @Bean
+    public Job indexingJob(JobRepository jobRepository, Step indexingStep) {
+        LOGGER.info("Chunk size: {}, Parallelization size: {}", chunkSize, parallelization);
+        return new JobBuilder(BATCH_JOB, jobRepository)
+                .incrementer(new TimestampJobParametersIncrementer())
+                .start(indexingStep)
+                .build();
+    }
+
+    @Bean
+    public Step indexingStep(
+            JobRepository jobRepository,
+            @Qualifier("transactionManager") PlatformTransactionManager transactionManager,
+            RepositoryItemReader<ExecutionRecord> indexingRepositoryItemReader,
+            ItemProcessor<ExecutionRecord, Future<ExecutionRecordDTO>> indexingAsyncItemProcessor,
+            ItemWriter<Future<ExecutionRecordDTO>> executionRecordDTOAsyncItemWriter) {
+
+        return new StepBuilder(STEP_NAME, jobRepository)
+                .<ExecutionRecord, Future<ExecutionRecordDTO>>chunk(chunkSize, transactionManager)
+                .reader(indexingRepositoryItemReader)
+                .processor(indexingAsyncItemProcessor)
+                .writer(executionRecordDTOAsyncItemWriter)
+                .build();
+    }
+
+    @Bean
+    @StepScope
+    public RepositoryItemReader<ExecutionRecord> indexingRepositoryItemReader(
+            ExecutionRecordRepository executionRecordRepository) {
+        final DefaultRepositoryItemReader defaultRepositoryItemReader = new DefaultRepositoryItemReader(executionRecordRepository);
+        defaultRepositoryItemReader.setPageSize(chunkSize);
+        return defaultRepositoryItemReader;
+    }
+
+    @Bean
+    public ItemProcessor<ExecutionRecord, Future<ExecutionRecordDTO>> indexingAsyncItemProcessor(
+            ItemProcessor<ExecutionRecord, ExecutionRecordDTO> indexingItemProcessor,
+            @Qualifier("indexingStepAsyncTaskExecutor") TaskExecutor indexingAsyncTaskExecutor) {
+        AsyncItemProcessor<ExecutionRecord, ExecutionRecordDTO> asyncItemProcessor = new AsyncItemProcessor<>();
+        asyncItemProcessor.setDelegate(indexingItemProcessor);
+        /*TODO:
```

Comment:
Where do you see that?
I was under the impression that what we have and setup for indexing is thread-safe.

Reply:
This is not true.
https://www.mongodb.com/docs/drivers/java/sync/v4.9/fundamentals/connection/connect/#:~:text=As%20each%20MongoClient%20represents%20a,driver%2C%20see%20the%20FAQ%20page.
The MongoClient is thread safe and provides a pool of connections.
Additionally, the IndexerFactory was created for that specific reason(for ecloud I believe), to allow multi-threads on all the clients that are necessary during indexing.
Why the threads do not close, that could be a bug and perhaps we should investigate further.
Do you want to tackle this here or should we create another ticket?

Reply:
I will spent more time on it here.

Reply:
You are right. I added task executor to processor. Job seems to work fine.